12/README.md

```diff
@@ -0,0 +1,189 @@
+
+*Tango Server Library* (server kernel library) - library that encapsulates *Tango Server* features and provides *Tango Server API* to a user.
+
+*Tango Server API* (server API) - an API that user uses to develop *Tango Device* functionality.
+
+*Upstream Tango Server/Device* (upstream server/device) - a *Tango Device* to which updates *client* subscribes for.  
+
+*Polling thread* (polling thread) - a thread within *Tango Server* that polls periodically defined *Tango Device*'s attributes/sometimes commands.
+
+*Tango Publish-Subscriber protocol* (pub-sub) - is a sub-set of *Tango Client-Server protocol* [RFC-?] that defines a relationship between *client* and *server* established via a negotiation phase and lasts till explicitly broken. During this relationship client receives updates published by *upstream server*. Updates are sourced from server's *polling thread* or directly through *server API*. Updates are filtered according to *device*'s configuration.
+
+*Tango Event* (event) - an update sent by *server*.
+
+*Tango Event Callback* (callback) - a piece of code that *client library* executes when an update is received.
+
+*Channel* - virtual client-server point-to-point connection for transmitting events
+
+## Definitions
+
+This section offers a number of ABNF formal definitions for entities repeatedly used below in this document.
+
+> **NOTE**: in the below ABNF some basic definitions are used as rules, spaces are replaced with '_' 
+
+> **NOTE**: in the below ABNF data types sometimes embeded into the rule e.g. bool:isExcept is equivalent to isExcept = true / false
+
+```ABNF    
+endpoint = URL ; usually with port
+
+SUBSCRIBER_INFO = upstream_device attribute_name action EVENT_TYPE client_idl_ver                                                                                                 
+                                                                
+SUBSCRIPTION_INFO = server_library_release_ver upstream_device_idl_ver zmq_sub_event_hwm rate ivl zmq_release_ver heartbeat_endpoint event_endpoint 1*alternate_heartbeat_endpoint 1*alternate_event_endpoint HEARTBEAT_CHANNEL EVENT_CHANNEL
+              
+EVENT_INFO = EVENT_DATA/1*EVENT_ERROR bool:isExcept  ; see below
+
+EVENT_DATA = ATT_CONF_DATA/PIPE_DATA/DATA_READY_DATA/INTERFACE_CHANGE_DATA/ATTRIBUTE_DATA 
+             
+ATT_CONF_DATA = ATTRIBUTE_PROPERTIES ; Attribute properties from RFC-4, see link below
+
+PIPE_DATA =  int:size name timeVal PIPE_BLOB ; Pipe blob must be defined in RFC-7
+
+DATA_READY_DATA = attribute_name data_type int:counter
+
+INTERFACE_CHANGE_DATA = *ATTRIBUTE_PROPERTIES *COMMAND_META_INFO bool:deviceStarted
+
+ATTRIBUTE_DATA = ATTRIBUTE_DEFINITION ; Attribute definition from RFC-4 
+
+EVENT_ERROR = reason severity desc origin
+
+EVENT_TYPE = "QUALITY_EVENT"/"INTERFACE_CHANGE"/"PIPE"/"ATTR_CONF_EVENT"/"CHANGE_EVENT"/"PERIODIC_EVENT"/"ARCHIVE_EVENT"/"USER_EVENT"/"HEARTBEAT"
```

Comment:
QUALITY_EVENT is commented as deprecated in cppTango:
https://github.com/tango-controls/cppTango/blob/40aecac1388b5190ec281d99659d170f35a2d5a2/cppapi/server/tango_const.h.in#L1106

```suggestion
EVENT_TYPE = "INTERFACE_CHANGE"/"PIPE_EVENT"/"ATTR_CONF_EVENT"/"CHANGE_EVENT"/"PERIODIC_EVENT"/"ARCHIVE_EVENT"/"USER_EVENT"/"HEARTBEAT"
```

Reply:
I don't think the Tango client API lets you subscribe to HEARTBEAT events directly.
This HEARTBEAT event subscription is done implicitly when subscribing to another event (it is a low level subscription).
HEARTBEAT as EVENT_TYPE is never used in the SUBSCRIBER_INFO which is listed above.
This is why HEARBEAT is not listed in cppTango enumType enumeration:
https://github.com/tango-controls/cppTango/blob/40aecac1388b5190ec281d99659d170f35a2d5a2/cppapi/server/tango_const.h.in#L1104-L1115